File: src/test/java/com/polygloat/controllers/ExportControllerTest.java
package com.polygloat.controllers;

import com.polygloat.model.Repository;
import org.springframework.test.web.servlet.MvcResult;
import org.testng.annotations.Test;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

public class ExportControllerTest extends SignedInControllerTest {

    @Test
    public void exportZipJson() throws Exception {
        Repository repository = dbPopulator.populate(generateUniqueString());
        commitTransaction();
        MvcResult mvcResult = performGet("/api/repository/" + repository.getId() + "/export/jsonZip")
                .andExpect(status().isOk()).andDo(MvcResult::getAsyncResult).andReturn();
        mvcResult.getResponse();
        Map<String, Long> fileSizes = parseZip(mvcResult.getResponse().getContentAsByteArray());


        repository.getLanguages().forEach(l -> {
            String name = l.getAbbreviation() + ".json";
            assertThat(fileSizes).containsKey(name);
            //assertThat(fileSizes.get(name)).isGreaterThan(0);
        });
        //cleanup
        repositoryService.deleteRepository(repository.getId());
    }


    private Map<String, Long> parseZip(byte[] responseContent) throws IOException {
        ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(responseContent);
        ZipInputStream zipInputStream = new ZipInputStream(byteArrayInputStream);

        HashMap<String, Long> result = new HashMap<>();

        ZipEntry nextEntry;
        while ((nextEntry = zipInputStream.getNextEntry()) != null) {
            result.put(nextEntry.getName(), nextEntry.getSize());
        }

        return result;
    }
}

File: src/test/java/io/tolgee/controllers/ExportControllerTest.kt
package io.tolgee.controllers

import io.tolgee.annotations.ProjectApiKeyAuthTestMethod
import io.tolgee.constants.ApiScope
import io.tolgee.fixtures.andIsForbidden
import io.tolgee.fixtures.generateUniqueString
import io.tolgee.model.Language
import org.assertj.core.api.Assertions
import org.springframework.test.web.servlet.MvcResult
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import org.springframework.transaction.annotation.Transactional
import org.testng.annotations.Test
import java.io.ByteArrayInputStream
import java.util.function.Consumer
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream

class ExportControllerTest : ProjectAuthControllerTest() {
    @Test
    @Transactional
    fun exportZipJson() {
        val project = dbPopulator.populate(generateUniqueString())
        commitTransaction()
        val mvcResult = performAuthGet("/api/project/" + project.id + "/export/jsonZip")
                .andExpect(MockMvcResultMatchers.status().isOk).andDo { obj: MvcResult -> obj.asyncResult }.andReturn()
        mvcResult.response
        val fileSizes = parseZip(mvcResult.response.contentAsByteArray)
        project.languages.forEach(Consumer { l: Language ->
            val name = l.abbreviation + ".json"
            Assertions.assertThat(fileSizes).containsKey(name)
        })
        //cleanup
        projectService.deleteRepository(project.id)
    }

    @Test
    @Transactional
    @ProjectApiKeyAuthTestMethod
    fun exportZipJsonWithApiKey() {
        projectSupplier = { dbPopulator.populate(generateUniqueString()).also { commitTransaction() } }
        val mvcResult = performRepositoryAuthGet("export/jsonZip")
                .andExpect(MockMvcResultMatchers.status().isOk).andDo { obj: MvcResult -> obj.asyncResult }.andReturn()
        mvcResult.response
        val fileSizes = parseZip(mvcResult.response.contentAsByteArray)
        project.languages.forEach(Consumer { l: Language ->
            val name = l.abbreviation + ".json"
            Assertions.assertThat(fileSizes).containsKey(name)
        })
        //cleanup
        projectService.deleteRepository(project.id)
    }

    @Test
    @ProjectApiKeyAuthTestMethod(scopes = [ApiScope.KEYS_EDIT])
    fun exportZipJsonApiKeyPermissionFail() {
        performRepositoryAuthGet("export/jsonZip").andIsForbidden
    }

    private fun parseZip(responseContent: ByteArray): Map<String, Long> {
        val byteArrayInputStream = ByteArrayInputStream(responseContent)
        val zipInputStream = ZipInputStream(byteArrayInputStream)
        val result = HashMap<String, Long>()
        var nextEntry: ZipEntry?
        while (zipInputStream.nextEntry.also {
                    nextEntry = it
                } != null) {
            result[nextEntry!!.name] = nextEntry!!.size
        }
        return result
    }
}
